### PR TITLE
#914: [typescript .d.ts] TextOverflow, Button, StatefulButton, Panel, PanelHeader, TabbedPanel and Tooltip (closes #914)

### LINKED ISSUE #914: [typescript .d.ts] Tooltip
← #908

## requirements
{describe the new feature}

## specs
{optional: describe technical specs to implement this feature, if not obvious}

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by ascariandrea**
Closes #914

## Test Plan

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.

- [ ] Google Chrome
- [ ] Internet Explorer

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

File: src/button/Button.d.ts [hunk 1]
```diff
@@ -21,9 +22,17 @@ export type ButtonProps = {
     fluid?: boolean,
     circular?: boolean,
     textOverflow?: React.ComponentClass<object>, // todo(typo) when translate text overflow, update here its props types
```
Comment (francescogior) [inline src/button/Button.d.ts]: also here ComponentType, and you may pass Textoverflow's props now

File: src/button/Button.d.ts [hunk 2]
```diff
@@ -21,9 +22,17 @@ export type ButtonProps = {
...
-    style?: React.CSSProperties,
+    style?: CSSProperties,
     className?: string
   }
```
Comment (francescogior) [inline src/button/Button.d.ts]: what about StatefulButtonProps?
they should be `ButtonProps & { baseState?: BaseState, stableSuccess?: boolean, timerMillis?: number }`

Comment (francescogior) [inline src/button/Button.d.ts]: also, there where tons of validations (`propsInvariants`), probably impossible to replicate in TS, but I'd add a TODO as usual

File: src/button/Button.d.ts [hunk 3]
```diff
@@ -21,9 +22,17 @@ export type ButtonProps = {
...
-export default class Button extends React.PureComponent<ButtonProps> {}
+export default class Button extends PureComponent<ButtonProps> {}
+
+export const Props: {
+  [key: string]: Type<any>
+}
+
+export const buttonStates: ButtonProps.ButtonState;
+export const buttonSizes: ButtonProps.ButtonSize;
+export const buttonTypes: ButtonProps.ButtonType;
```
Comment (francescogior) [inline src/button/Button.d.ts]: the three above are arrays

```js
export const buttonStates: ButtonProps.ButtonState[];
export const buttonSizes: ButtonProps.ButtonSize[];
export const buttonTypes: ButtonProps.ButtonType[];
```

File: src/Panel/Panel.d.ts [hunk 4]
```diff
@@ -35,3 +36,7 @@ export type PanelProps = {
 }
```
Comment (francescogior) [inline src/Panel/Panel.d.ts]: why `docked-top' | ...` is not declared in the namespace like the others?

File: src/Tooltip/Tooltip.d.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import { PopoverProps } from '../popover/Popover';
+
+export namespace TooltipProps {
+  type Type = 'light' | 'dark';
+  type Size = 'small' | 'big';
+}
+
+export type TooltipProps = {
+  children: any, // TODO: t.ReactChildren
+  popover: PopoverProps,
```
Comment (francescogior) [inline src/Tooltip/Tooltip.d.ts]: add TODO, was refinement 
`p => t.String.is(p.content) && t.Nil.is(p.event))`

File: src/text-overflow/TextOverflow.d.ts [hunk 6]
```diff
@@ -1,13 +1,18 @@
 import { CSSProperties, PureComponent } from 'react';
+import { Type } from 'tcomb';
 
 export type TextOverflowProps = {
   children?: (self: JSX.Element, isOpen: boolean) => any,
   label: string,
+  popover?: object, // TODO(typo): Omit?
```
Comment (FrancescoCioria) [inline src/text-overflow/TextOverflow.d.ts]: @ascariandrea as I told you in the other PRs, this should reflect exactly the `@props` defined in this repo.

Please don't copy-paste from alinity-pro, those are the props of the custom wrappers.
If you start from those you will always make mistakes as they are a different thing :)

Comment (ascariandrea) [inline src/text-overflow/TextOverflow.d.ts]: I think I made a wrong rebase :/

File: src/button/Button.d.ts (hunk 2 above)
Comment (ascariandrea) [inline src/button/Button.d.ts]: > what about StatefulButtonProps?

They're defined in `StatefulButton.d.ts`

File: src/Panel/Panel.d.ts (hunk 4 above)
Comment (ascariandrea) [inline src/Panel/Panel.d.ts]: I've changed the exports to replicate the originals

**Pull request merged**